### PR TITLE
Reorganize Util classes (with BC breaks)

### PR DESCRIPTION
**Pull request by Ener-Getick**
In this PR, I rewrote the ``FormatNegotiator`` to use willdurant/negotiation FormatNegotiatorInterface and the requestStack.
I also remove the class/methods deprecated in #1095.

Only the last commit is relevant; the others are from #1095.

### CODE REVIEW

**Comment by lsmith77** [comment]
please rebase on master

**Comment by lsmith77** [comment]
though you might also want to wait on #1104

**Comment by Ener-Getick** [comment]
As you want, it doesn't change anything.

File: EventListener/FormatListener.php [hunk 1]
```diff
@@ -51,15 +50,15 @@ public function onKernelRequest(GetResponseEvent $event)
...
             $format = $request->getRequestFormat(null);
             if (null === $format) {
-                if ($this->formatNegotiator instanceof MediaTypeNegotiatorInterface) {
-                    $mediaType = $this->formatNegotiator->getBestMediaType($request);
-                    if ($mediaType) {
-                        $request->attributes->set('media_type', $mediaType);
-                        $format = $request->getFormat($mediaType);
-                    }
-                } else {
+                // if ($this->formatNegotiator instanceof MediaTypeNegotiatorInterface) {
```
Comment (lsmith77) [inline EventListener/FormatListener.php]: just removed outdated code

**Comment by lsmith77** [comment]
yeah I also just noticed that #1104 only focuses on ParamFetcher ..

anyways .. looks like this needs a rebase again ..

File: EventListener/FormatListener.php (hunk 1 above)
Comment (Ener-Getick) [inline EventListener/FormatListener.php]: I have to adapt this code but I don't know how work the version listener...

**Comment by Ener-Getick** [comment]
We must change the minimum version of willdurand/negotiation if we want to use the ``FormatNegotiationInterface``. This interface only appear with the 1.3.0 ...

File: EventListener/FormatListener.php (hunk 1 above)
Comment (lsmith77) [inline EventListener/FormatListener.php]: yeah .. it tries to parse a version string from the mime type

**Comment by Ener-Getick** [comment]
@willdurand said that willdurand/negotiation@2.0 will be released as stable in a few weeks (https://github.com/willdurand/Negotiation/issues/56#issuecomment-127066087).

I think it could be a good idea to update this dependency version, no ?

File: Controller/ExceptionController.php [hunk 2]
```diff
@@ -195,7 +195,10 @@ protected function getFormat(Request $request, $format)
     {
         try {
             $formatNegotiator = $this->container->get('fos_rest.exception_format_negotiator');
-            $format = $formatNegotiator->getBestFormat($request) ?: $format;
+            $accept = $formatNegotiator->getBest('', array());
```
Comment (amenophis) [inline Controller/ExceptionController.php]: You could put this line in the if below ?

Comment (Ener-Getick) [inline Controller/ExceptionController.php]: Is that a joke?

Comment (amenophis) [inline Controller/ExceptionController.php]: not really, why ?

Comment (Ener-Getick) [inline Controller/ExceptionController.php]: This if checks $accept value, so if I move this line which defines $accept, how could the condition be true?...

**Comment by lsmith77** [comment]
needs a rebase

**Comment by Ener-Getick** [comment]
done.

File: Controller/ExceptionController.php [hunk 3]
```diff
@@ -195,7 +195,10 @@ protected function getFormat(Request $request, $format)
...
+            if($accept) {
```
Comment (lsmith77) [inline Controller/ExceptionController.php]: space missing

File: EventListener/FormatListener.php [hunk 4]
```diff
@@ -51,14 +50,12 @@ public function onKernelRequest(GetResponseEvent $event)
...
-                if ($this->formatNegotiator instanceof MediaTypeNegotiatorInterface) {
-                    $mediaType = $this->formatNegotiator->getBestMediaType($request);
-                    if (null !== $mediaType) {
-                        $request->attributes->set('media_type', $mediaType);
-                        $format = $request->getFormat($mediaType);
-                    }
-                } else {
-                    $format = $this->formatNegotiator->getBestFormat($request);
+                $accept = $this->formatNegotiator->getBest('', array());
+                if (null !== $accept && 0.0 < $accept->getQuality() &&
```
Comment (lsmith77) [inline EventListener/FormatListener.php]: I think the above code is a bit hard to read .. maybe this is better ..

```
if (null !== $accept && 0.0 < $accept->getQuality()) {
    $format = $request->getFormat($accept->getType();
    if (null !== $format) {
        $request->attributes->set('media_type', $accept->getValue());
    }
}

File: EventListener/MimeTypeListener.php [hunk 5]
```diff
@@ -50,7 +46,6 @@ public function onKernelRequest(GetResponseEvent $event)
         if (HttpKernelInterface::MASTER_REQUEST === $event->getRequestType()) {
             foreach ($this->mimeTypes['formats'] as $format => $mimeType) {
                 $request->setFormat($format, $mimeType);
-                $this->formatNegotiator->registerFormat($format, (array) $mimeType, true);
```
Comment (lsmith77) [inline EventListener/MimeTypeListener.php]: this code was necessary to explicitly register supported mime types .. is this really no longer necessary?

Comment (Ener-Getick) [inline EventListener/MimeTypeListener.php]: This is no longer necessary because I now use the ``Request`` to find the format corresponding to a mime type.

**Comment by lsmith77** [comment]
awesome!

File: Negotiation/FormatNegotiator.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                    if ($header && $extensionHeader) {
+                        $extensionHeader = ','.$extensionHeader;
+                    }
+                }
+                if ($extensionHeader) {
+                    $header .= $extensionHeader.'; q='.$options['prefer_extension'];
+                }
+            }
+
+            $mimeTypes = $this->normalizePriorities(empty($priorities) ? $options['priorities'] : $priorities);
```
Comment (lsmith77) [inline Negotiation/FormatNegotiator.php]: this method no longer exists in 2.0

Comment (lsmith77) [inline Negotiation/FormatNegotiator.php]: @willdurand I am not sure how we can revive this anyway .. the array with formats to mime types mapping is protected in the ``Request`` class .. so there is no way to nicely "explode" a format to an array of enabled mime types, so I guess users will just need to specify mime types from now on? or we just map a format to the "primary" mime type

Comment (lsmith77) [inline Negotiation/FormatNegotiator.php]: @Ener-Getick it seems like we have no tests for this code path :(

Comment (Ener-Getick) [inline Negotiation/FormatNegotiator.php]: Hum, the solution may be to copy the ``Request::$formats``array... Or as you said, we can also force the users to only provide mime types.

Comment (Ener-Getick) [inline Negotiation/FormatNegotiator.php]: I created a new PR https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1149.

**Pull request merged**